### ISSUE TITLE
Invalid refresherToken reAuthenticate never triggered

### ISSUE THREAD

**Issue by MB175** [cloudfoundry-community/go-cfclient#345]
The current implementation of the request executor leads to an error if the refresherToken is invalid or expiring.

Manually reproduced, the log can vary, for example if the refreshToken is expired. 
`error executing request, failed during HTTP request send: Get "...": oauth2: "invalid_token" "The token expired, was revoked, or the token ID is incorrect."`

As far as I understand, there is a check for status code 401 that is meant to indicate that the used token is invalid and properly didn't get renewed by the refresherToken.

https://github.com/cloudfoundry-community/go-cfclient/blob/7955029538e46e03a287cc475f8ca4ae02e754cd/internal/http/executor.go#L56

However, this is not working out. If the refresher token is invalid or expired, the if condition in line 56 is never reached since the error from the underlying oauth logic is passed forward and is already handled way before Line 51

https://github.com/cloudfoundry-community/go-cfclient/blob/7955029538e46e03a287cc475f8ca4ae02e754cd/internal/http/executor.go#L51

Furthermore, there is no Statuscode passed because if the error occurs, nil is passed instead of the request object

https://github.com/cloudfoundry-community/go-cfclient/blob/7955029538e46e03a287cc475f8ca4ae02e754cd/internal/http/executor.go#L121

---

### Steps to reproduce
If you don't have control over the lifetime of the tokens, there is a simple way to manually reproduce this issue.

- Replace the "RefreshToken" token with an empty value with the help of a debugger, this can be done, for example in the `golang.org/x/oauth2/internal/token.go` module (line 293 RefreshToken)
- Wait for the actual AccessToken to expire
- The Issue will appear when a call is made after the AccessToken is expired and the client trys to renew it (best observed white a break point at the Executor Line 51)

https://github.com/cloudfoundry-community/go-cfclient/blob/7955029538e46e03a287cc475f8ca4ae02e754cd/internal/http/executor.go#L51

**Comment by MB175** [comment]
Extending the error check like that, should properly do the trick:

```GO
	// do the request to the remote API
	r, err := c.do(req, followRedirects)
	if err != nil {
		var oauthErr *oauth2.RetrieveError
		if errors.As(err, &oauthErr) {
			if oauthErr.Response.StatusCode == http.StatusUnauthorized {
				err = c.reAuthenticate()
				if err != nil {
					return nil, err
				}
				r, err = c.do(req, followRedirects)
			}
		}
		return nil, err
	}

```

**Comment by MB175** [comment]
> Extending the error check like that, should properly do the trick:
> 
> 
> 
> ```GO
> 
> 	// do the request to the remote API
> 
> 	r, err := c.do(req, followRedirects)
> 
> 	if err != nil {
> 
> 		var oauthErr *oauth2.RetrieveError
> 
> 		if errors.As(err, &oauthErr) {
> 
> 			if oauthErr.Response.StatusCode == http.StatusUnauthorized {
> 
> 				err = c.reAuthenticate()
> 
> 				if err != nil {
> 
> 					return nil, err
> 
> 				}
> 
> 				r, err = c.do(req, followRedirects)
> 
> 			}
> 
> 		}
> 
> 		return nil, err
> 
> 	}
> 
> 
> 
> ```

As I'm thinking about that we need to go one function deeper, since the error is already wrapped again here. So we can't cast it.

**Comment by sneal** [comment]
How are you providing credentials? Are you using `config.NewFromCFHomeDir()` or something else?

**Comment by sneal** [comment]
Fixed in 2eff8ed70e0689599392490966c4eff86ed2a9c8
